# FLSim Tutorial: Features to compress client updates for communication efficiency


## Introduction

In this tutorial, we will look at how to use FL Simulator channels to compress client to server messages for communication efficiency. Several approaches, such as quantization, sparsity and sketching have been shown to be useful, and here we guide you through using support in the simulator for three basic techniques, namely half-precision, scalar quantization and Top-K sparsity.

### Prerequisites

Communication efficiency in federated learning is an important problem, as we have to transmit messages between the server (which is usually in a datacenter) and clients, which for example could be mobile devices communicating through a low-bandwidth channel. This causes a bottleneck, specially for the download phase where clients upload model updates to the server. One of the ways this can be addressed is by applying different techniques for compressing the client update. To get started with the motivation for this and get introduced to some of the background techniques, here is a recommended reading list:
- Jakub Konečný, H. Brendan McMahan, Felix X. Yu, Peter Richtarik, Ananda Theertha Suresh, Dave Bacon, "Federated Learning: Strategies for Improving Communication Efficiency", 
NIPS Workshop on Private Multi-Party Machine Learning (2016) https://research.google/pubs/pub45648/
- Sattler, Felix, et al. "Robust and communication-efficient federated learning from non-iid data." IEEE transactions on neural networks and learning systems 31.9 (2019): 3400-3413. https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=8889996


Now that you're familiar with the motivation and basic techniques for communication efficiency, let's get started!


### Objectives 

By the end of this tutorial, we will have learnt how to

1. Setup a basic model training pipeline on the Cifar10 task
2. Configure different channels for quantization and sparsity
3. Setup channel measurements for tracking communication
5. Launch FL training flows using FLSim with quantization and sparsity

## Setup a basic training pipeline for the Cifar10 task

### Prerequisite
Let's install flsim via pip with the command below.

In [38]:
!pip install --quiet flsim

Subsequently, let us define the data pipeline and model for training on the Cifar10 task. This is similar to what we have done in the Cifar-10 tutorial at https://github.com/facebookresearch/FLSim/blob/main/tutorials/cifar10_tutorial.ipynb. 

In [39]:
USE_CUDA = True
LOCAL_BATCH_SIZE = 32
EXAMPLES_PER_USER = 500
IMAGE_SIZE = 32

We set up the data pipeline for this task and import the model we will be training on Cifar-10. Similar to the Cifar-10 tutorial, it will be a standard four-layer convolutional neural network, with group norms and a linear layer. We also import the metrics reporter as in the original tutorial.

In [40]:
import torch
from torchvision.datasets.cifar import CIFAR10
from flsim.utils.example_utils import FLModel
from flsim.utils.example_utils import SimpleConvNet

cuda_enabled = torch.cuda.is_available() and USE_CUDA
device = torch.device(f"cuda:{0}" if cuda_enabled else "cpu")

We define a config for training on the Cifar-10 dataset with federated learning. This does not incorporate any settings for client update compression yet, as we will be showing in subsequent sections how to choose a technique and define the configuration for it. We will be modifying this config to incorporate different approaches such as half-precision, scalar quantization and sparsity, and also guide you through their compression parameters.

In [41]:
import flsim.configs
from flsim.utils.config_utils import fl_config_from_json
from omegaconf import OmegaConf


json_config = {
    "trainer": {
        "_base_": "base_sync_trainer",
        # there are different types of aggregators
        # fed avg doesn't require lr, while others such as fed_avg_with_lr or fed_adam do
        "_base_": "base_sync_trainer",
        "server": {
          "_base_": "base_sync_server",
          "server_optimizer": {
            "_base_": "base_fed_avg_with_lr",
            "lr": 2.13,
            "momentum": 0.9
          },
          # type of user selection sampling
          "active_user_selector": {"_base_": "base_uniformly_random_active_user_selector"},
        },
        "client": {
            # number of client's local epoch
            "epochs": 1,
            "optimizer": {
                "_base_": "base_optimizer_sgd",
                # client's local learning rate
                "lr": 0.0138,
                # client's local momentum
                "momentum": 0,
            },
        },
        # number of users per round for aggregation
        "users_per_round": 5,
        # total number of global epochs
        # total #rounds = ceil(total_users / users_per_round) * epochs
        "epochs": 1,
        # frequency of reporting train metrics
        "train_metrics_reported_per_epoch": 100,
        # frequency of evaluation per epoch
        "eval_epoch_frequency": 1,
        "do_eval": True,
        # should we report train metrics after global aggregation
        "report_train_metrics_after_aggregation": True,
    }
}

We now define a function *run_cifar10_fl_training()* which runs a round of FL training, and takes as input the FL configuration. We can train under different compression techniques by appropriately changing the config as needed.

In [42]:
from flsim.utils.example_utils import build_data_provider, MetricsReporter

data_provider = build_data_provider(
    local_batch_size=LOCAL_BATCH_SIZE,
    examples_per_user=EXAMPLES_PER_USER,
    image_size = IMAGE_SIZE,
)

from flsim.interfaces.metrics_reporter import Channel

from hydra.utils import instantiate

def run_cifar10_fl_training(json_config):
  model = SimpleConvNet(in_channels=3, num_classes=10)
  global_model = FLModel(model, device)

  if cuda_enabled:
      global_model.fl_cuda()

  cfg = fl_config_from_json(json_config)
  trainer_config = cfg.trainer
  trainer = instantiate(trainer_config, model=global_model, cuda_enabled=cuda_enabled)

  metrics_reporter = MetricsReporter([Channel.TENSORBOARD, Channel.STDOUT])

  final_model, eval_score = trainer.train(
      data_provider=data_provider,
      metrics_reporter=metrics_reporter,
      num_total_users=data_provider.num_train_users(),
      distributed_world_size=1
  )

Files already downloaded and verified
Files already downloaded and verified


Creating FL User: 100user [00:24,  4.11user/s]
Creating FL User: 20user [00:03,  5.33user/s]
Creating FL User: 20user [00:03,  5.37user/s]

Clients in total: 100


Let us first train a vanilla setting where there is no client update compression, and no metrics reporting. 

In [43]:
run_cifar10_fl_training(json_config)

/usr/local/lib/python3.7/dist-packages/hydra/experimental/initialize.py:36: UserWarning: hydra.experimental.initialize() is no longer experimental. Use hydra.initialize()
  message="hydra.experimental.initialize() is no longer experimental."
/usr/local/lib/python3.7/dist-packages/hydra/experimental/compose.py:19: UserWarning: hydra.experimental.compose() is no longer experimental. Use hydra.compose()
  message="hydra.experimental.compose() is no longer experimental."
Round:   0%|          | 0/20 [00:00<?, ?round/s]

Train finished Global Round: 1
(epoch = 1, round = 1, global round = 1), Loss/Training: 2.312931588292122
(epoch = 1, round = 1, global round = 1), Accuracy/Training: 13.88
reporting (epoch = 1, round = 1, global round = 1) for aggregation


Round:   5%|▌         | 1/20 [00:01<00:26,  1.42s/round]

(epoch = 1, round = 1, global round = 1), Loss/Aggregation: 2.1286049485206604
(epoch = 1, round = 1, global round = 1), Accuracy/Aggregation: 27.08
Train finished Global Round: 2
(epoch = 1, round = 2, global round = 2), Loss/Training: 2.1240725725889207
(epoch = 1, round = 2, global round = 2), Accuracy/Training: 24.12
reporting (epoch = 1, round = 2, global round = 2) for aggregation


Round:  10%|█         | 2/20 [00:03<00:27,  1.53s/round]

(epoch = 1, round = 2, global round = 2), Loss/Aggregation: 1.9662387758493423
(epoch = 1, round = 2, global round = 2), Accuracy/Aggregation: 29.96
Train finished Global Round: 3
(epoch = 1, round = 3, global round = 3), Loss/Training: 1.964783202111721
(epoch = 1, round = 3, global round = 3), Accuracy/Training: 29.44
reporting (epoch = 1, round = 3, global round = 3) for aggregation


Round:  15%|█▌        | 3/20 [00:04<00:26,  1.58s/round]

(epoch = 1, round = 3, global round = 3), Loss/Aggregation: 1.9338369429111482
(epoch = 1, round = 3, global round = 3), Accuracy/Aggregation: 28.48
Train finished Global Round: 4
(epoch = 1, round = 4, global round = 4), Loss/Training: 1.7915167212486267
(epoch = 1, round = 4, global round = 4), Accuracy/Training: 36.12
reporting (epoch = 1, round = 4, global round = 4) for aggregation


Round:  20%|██        | 4/20 [00:06<00:26,  1.66s/round]

(epoch = 1, round = 4, global round = 4), Loss/Aggregation: 1.6435432597994803
(epoch = 1, round = 4, global round = 4), Accuracy/Aggregation: 41.76
Train finished Global Round: 5
(epoch = 1, round = 5, global round = 5), Loss/Training: 1.7010600984096527
(epoch = 1, round = 5, global round = 5), Accuracy/Training: 39.32
reporting (epoch = 1, round = 5, global round = 5) for aggregation


Round:  25%|██▌       | 5/20 [00:08<00:24,  1.66s/round]

(epoch = 1, round = 5, global round = 5), Loss/Aggregation: 1.6652007445693016
(epoch = 1, round = 5, global round = 5), Accuracy/Aggregation: 41.08
Train finished Global Round: 6
(epoch = 1, round = 6, global round = 6), Loss/Training: 1.6321644321084023
(epoch = 1, round = 6, global round = 6), Accuracy/Training: 42.16
reporting (epoch = 1, round = 6, global round = 6) for aggregation


Round:  30%|███       | 6/20 [00:09<00:22,  1.60s/round]

(epoch = 1, round = 6, global round = 6), Loss/Aggregation: 1.5614833593368531
(epoch = 1, round = 6, global round = 6), Accuracy/Aggregation: 43.76
Train finished Global Round: 7
(epoch = 1, round = 7, global round = 7), Loss/Training: 1.5849987655878066
(epoch = 1, round = 7, global round = 7), Accuracy/Training: 41.24
reporting (epoch = 1, round = 7, global round = 7) for aggregation


Round:  35%|███▌      | 7/20 [00:10<00:18,  1.42s/round]

(epoch = 1, round = 7, global round = 7), Loss/Aggregation: 1.5697927251458168
(epoch = 1, round = 7, global round = 7), Accuracy/Aggregation: 43.64
Train finished Global Round: 8
(epoch = 1, round = 8, global round = 8), Loss/Training: 1.5407207921147346
(epoch = 1, round = 8, global round = 8), Accuracy/Training: 45.32
reporting (epoch = 1, round = 8, global round = 8) for aggregation


Round:  40%|████      | 8/20 [00:11<00:15,  1.31s/round]

(epoch = 1, round = 8, global round = 8), Loss/Aggregation: 1.4618546932935714
(epoch = 1, round = 8, global round = 8), Accuracy/Aggregation: 47.72
Train finished Global Round: 9
(epoch = 1, round = 9, global round = 9), Loss/Training: 1.5054958000779153
(epoch = 1, round = 9, global round = 9), Accuracy/Training: 46.04
reporting (epoch = 1, round = 9, global round = 9) for aggregation


Round:  45%|████▌     | 9/20 [00:12<00:13,  1.24s/round]

(epoch = 1, round = 9, global round = 9), Loss/Aggregation: 1.406162577867508
(epoch = 1, round = 9, global round = 9), Accuracy/Aggregation: 49.88
Train finished Global Round: 10
(epoch = 1, round = 10, global round = 10), Loss/Training: 1.4426088862121105
(epoch = 1, round = 10, global round = 10), Accuracy/Training: 47.32
reporting (epoch = 1, round = 10, global round = 10) for aggregation


Round:  50%|█████     | 10/20 [00:13<00:11,  1.18s/round]

(epoch = 1, round = 10, global round = 10), Loss/Aggregation: 1.4564439199864865
(epoch = 1, round = 10, global round = 10), Accuracy/Aggregation: 46.8
Train finished Global Round: 11
(epoch = 1, round = 11, global round = 11), Loss/Training: 1.4529921159148216
(epoch = 1, round = 11, global round = 11), Accuracy/Training: 48.12
reporting (epoch = 1, round = 11, global round = 11) for aggregation


Round:  55%|█████▌    | 11/20 [00:14<00:10,  1.15s/round]

(epoch = 1, round = 11, global round = 11), Loss/Aggregation: 1.4125294663012027
(epoch = 1, round = 11, global round = 11), Accuracy/Aggregation: 49.12
Train finished Global Round: 12
(epoch = 1, round = 12, global round = 12), Loss/Training: 1.3679302617907525
(epoch = 1, round = 12, global round = 12), Accuracy/Training: 50.56
reporting (epoch = 1, round = 12, global round = 12) for aggregation


Round:  60%|██████    | 12/20 [00:15<00:08,  1.12s/round]

(epoch = 1, round = 12, global round = 12), Loss/Aggregation: 1.4171042308211326
(epoch = 1, round = 12, global round = 12), Accuracy/Aggregation: 49.84
Train finished Global Round: 13
(epoch = 1, round = 13, global round = 13), Loss/Training: 1.3944446809589863
(epoch = 1, round = 13, global round = 13), Accuracy/Training: 51.44
reporting (epoch = 1, round = 13, global round = 13) for aggregation


Round:  65%|██████▌   | 13/20 [00:17<00:07,  1.10s/round]

(epoch = 1, round = 13, global round = 13), Loss/Aggregation: 1.327131749689579
(epoch = 1, round = 13, global round = 13), Accuracy/Aggregation: 52.68
Train finished Global Round: 14
(epoch = 1, round = 14, global round = 14), Loss/Training: 1.3676495552062988
(epoch = 1, round = 14, global round = 14), Accuracy/Training: 50.96
reporting (epoch = 1, round = 14, global round = 14) for aggregation


Round:  70%|███████   | 14/20 [00:18<00:06,  1.09s/round]

(epoch = 1, round = 14, global round = 14), Loss/Aggregation: 1.5048811189830302
(epoch = 1, round = 14, global round = 14), Accuracy/Aggregation: 46.32
Train finished Global Round: 15
(epoch = 1, round = 15, global round = 15), Loss/Training: 1.3665127858519555
(epoch = 1, round = 15, global round = 15), Accuracy/Training: 51.44
reporting (epoch = 1, round = 15, global round = 15) for aggregation


Round:  75%|███████▌  | 15/20 [00:19<00:05,  1.09s/round]

(epoch = 1, round = 15, global round = 15), Loss/Aggregation: 1.3110274441540242
(epoch = 1, round = 15, global round = 15), Accuracy/Aggregation: 53.68
Train finished Global Round: 16
(epoch = 1, round = 16, global round = 16), Loss/Training: 1.324099837988615
(epoch = 1, round = 16, global round = 16), Accuracy/Training: 53.64
reporting (epoch = 1, round = 16, global round = 16) for aggregation


Round:  80%|████████  | 16/20 [00:20<00:04,  1.08s/round]

(epoch = 1, round = 16, global round = 16), Loss/Aggregation: 1.3443891003727912
(epoch = 1, round = 16, global round = 16), Accuracy/Aggregation: 53.28
Train finished Global Round: 17
(epoch = 1, round = 17, global round = 17), Loss/Training: 1.277423980087042
(epoch = 1, round = 17, global round = 17), Accuracy/Training: 54.52
reporting (epoch = 1, round = 17, global round = 17) for aggregation


Round:  85%|████████▌ | 17/20 [00:21<00:03,  1.07s/round]

(epoch = 1, round = 17, global round = 17), Loss/Aggregation: 1.2762474924325944
(epoch = 1, round = 17, global round = 17), Accuracy/Aggregation: 53.84
Train finished Global Round: 18
(epoch = 1, round = 18, global round = 18), Loss/Training: 1.291584499925375
(epoch = 1, round = 18, global round = 18), Accuracy/Training: 54.04
reporting (epoch = 1, round = 18, global round = 18) for aggregation


Round:  90%|█████████ | 18/20 [00:22<00:02,  1.07s/round]

(epoch = 1, round = 18, global round = 18), Loss/Aggregation: 1.3562731005251407
(epoch = 1, round = 18, global round = 18), Accuracy/Aggregation: 51.28
Train finished Global Round: 19
(epoch = 1, round = 19, global round = 19), Loss/Training: 1.2776811376214028
(epoch = 1, round = 19, global round = 19), Accuracy/Training: 54.84
reporting (epoch = 1, round = 19, global round = 19) for aggregation


Round:  95%|█████████▌| 19/20 [00:23<00:01,  1.07s/round]

(epoch = 1, round = 19, global round = 19), Loss/Aggregation: 1.2884995967149735
(epoch = 1, round = 19, global round = 19), Accuracy/Aggregation: 53.2
Train finished Global Round: 20
(epoch = 1, round = 20, global round = 20), Loss/Training: 1.2883846752345562
(epoch = 1, round = 20, global round = 20), Accuracy/Training: 53.64
reporting (epoch = 1, round = 20, global round = 20) for aggregation
(epoch = 1, round = 20, global round = 20), Loss/Aggregation: 1.320712050050497
(epoch = 1, round = 20, global round = 20), Accuracy/Aggregation: 52.08
Running (epoch = 1, round = 20, global round = 20) for Eval
(epoch = 1, round = 20, global round = 20), Loss/Eval: 1.3698589166626334
(epoch = 1, round = 20, global round = 20), Accuracy/Eval: 51.25


Epoch:   0%|          | 0/1 [00:25<?, ?epoch/s]


## Introduction to communication metrics reporting in FL Simulator

We support tracking of communication metrics through the flag *report_communication_metrics*, and for every round report the mean number of bytes transmitted both for uploads (client to server communication), as well as downloads (server to client communication). So far our focus is mainly on optimizing for upload efficiency as uploads are generally slower. Let us modify the JSON config above to add a *channel* key to specify channel subconfigs. This controls selection of the compression method, as well as metrics tracking. When *report_communication_metrics* is enabled, the mean upload and download bytes (averaged across participating clients in a round) are reported. Let us consider the simplest compression supported in the simulator, namely the [half-precision channel](https://github.com/facebookresearch/FLSim/blob/main/channels/half_precision_channel.py). Here, client updates (which by default are 32-bit floating points) are compressed to 16-bit. We can specify the compression method with the *_base_* key here in the sub-config.

In [44]:
json_config["trainer"]["channel"] = {
  "report_communication_metrics": True,
  "_base_": "base_half_precision_channel"
}
cfg = fl_config_from_json(json_config)


/usr/local/lib/python3.7/dist-packages/hydra/experimental/initialize.py:36: UserWarning: hydra.experimental.initialize() is no longer experimental. Use hydra.initialize()
  message="hydra.experimental.initialize() is no longer experimental."
/usr/local/lib/python3.7/dist-packages/hydra/experimental/compose.py:19: UserWarning: hydra.experimental.compose() is no longer experimental. Use hydra.compose()
  message="hydra.experimental.compose() is no longer experimental."


After defining the JSON config and converting to an FL config, we instantiate the trainer and start training with half-precision compression. While the server sends full-precision models to clients, the client uploads consume half the number of bytes as expected.

In [45]:
run_cifar10_fl_training(json_config)

/usr/local/lib/python3.7/dist-packages/hydra/experimental/initialize.py:36: UserWarning: hydra.experimental.initialize() is no longer experimental. Use hydra.initialize()
  message="hydra.experimental.initialize() is no longer experimental."
/usr/local/lib/python3.7/dist-packages/hydra/experimental/compose.py:19: UserWarning: hydra.experimental.compose() is no longer experimental. Use hydra.compose()
  message="hydra.experimental.compose() is no longer experimental."
Round:   0%|          | 0/20 [00:00<?, ?round/s]

Train finished Global Round: 1
(epoch = 1, round = 1, global round = 1), Loss/Training: 2.3368776977062224
(epoch = 1, round = 1, global round = 1), Accuracy/Training: 13.68
reporting (epoch = 1, round = 1, global round = 1) for aggregation


Round:   5%|▌         | 1/20 [00:01<00:20,  1.09s/round]

(epoch = 1, round = 1, global round = 1), Loss/Aggregation: 2.1055475905537606
(epoch = 1, round = 1, global round = 1), Accuracy/Aggregation: 24.68
(epoch = 1, round = 1, global round = 1), Client to Server Bytes Sent/Training: 63188.0
(epoch = 1, round = 1, global round = 1), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 2
(epoch = 1, round = 2, global round = 2), Loss/Training: 2.1054040387272837
(epoch = 1, round = 2, global round = 2), Accuracy/Training: 23.28
reporting (epoch = 1, round = 2, global round = 2) for aggregation


Round:  10%|█         | 2/20 [00:02<00:19,  1.09s/round]

(epoch = 1, round = 2, global round = 2), Loss/Aggregation: 1.9819558918476106
(epoch = 1, round = 2, global round = 2), Accuracy/Aggregation: 29.36
(epoch = 1, round = 2, global round = 2), Client to Server Bytes Sent/Training: 63188.0
(epoch = 1, round = 2, global round = 2), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 3
(epoch = 1, round = 3, global round = 3), Loss/Training: 1.9592920973896981
(epoch = 1, round = 3, global round = 3), Accuracy/Training: 29.8
reporting (epoch = 1, round = 3, global round = 3) for aggregation


Round:  15%|█▌        | 3/20 [00:03<00:18,  1.09s/round]

(epoch = 1, round = 3, global round = 3), Loss/Aggregation: 1.9140223771333695
(epoch = 1, round = 3, global round = 3), Accuracy/Aggregation: 30.24
(epoch = 1, round = 3, global round = 3), Client to Server Bytes Sent/Training: 63188.0
(epoch = 1, round = 3, global round = 3), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 4
(epoch = 1, round = 4, global round = 4), Loss/Training: 1.826326185464859
(epoch = 1, round = 4, global round = 4), Accuracy/Training: 34.04
reporting (epoch = 1, round = 4, global round = 4) for aggregation


Round:  20%|██        | 4/20 [00:04<00:17,  1.08s/round]

(epoch = 1, round = 4, global round = 4), Loss/Aggregation: 1.8746705412864686
(epoch = 1, round = 4, global round = 4), Accuracy/Aggregation: 33.72
(epoch = 1, round = 4, global round = 4), Client to Server Bytes Sent/Training: 63188.0
(epoch = 1, round = 4, global round = 4), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 5
(epoch = 1, round = 5, global round = 5), Loss/Training: 1.7053318053483963
(epoch = 1, round = 5, global round = 5), Accuracy/Training: 38.4
reporting (epoch = 1, round = 5, global round = 5) for aggregation


Round:  25%|██▌       | 5/20 [00:05<00:16,  1.08s/round]

(epoch = 1, round = 5, global round = 5), Loss/Aggregation: 1.6819366380572318
(epoch = 1, round = 5, global round = 5), Accuracy/Aggregation: 39.32
(epoch = 1, round = 5, global round = 5), Client to Server Bytes Sent/Training: 63188.0
(epoch = 1, round = 5, global round = 5), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 6
(epoch = 1, round = 6, global round = 6), Loss/Training: 1.6209901988506317
(epoch = 1, round = 6, global round = 6), Accuracy/Training: 42.0
reporting (epoch = 1, round = 6, global round = 6) for aggregation


Round:  30%|███       | 6/20 [00:06<00:16,  1.17s/round]

(epoch = 1, round = 6, global round = 6), Loss/Aggregation: 1.6821082472801208
(epoch = 1, round = 6, global round = 6), Accuracy/Aggregation: 42.32
(epoch = 1, round = 6, global round = 6), Client to Server Bytes Sent/Training: 63188.0
(epoch = 1, round = 6, global round = 6), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 7
(epoch = 1, round = 7, global round = 7), Loss/Training: 1.5624877616763115
(epoch = 1, round = 7, global round = 7), Accuracy/Training: 43.8
reporting (epoch = 1, round = 7, global round = 7) for aggregation


Round:  35%|███▌      | 7/20 [00:07<00:14,  1.15s/round]

(epoch = 1, round = 7, global round = 7), Loss/Aggregation: 1.7349697917699813
(epoch = 1, round = 7, global round = 7), Accuracy/Aggregation: 40.0
(epoch = 1, round = 7, global round = 7), Client to Server Bytes Sent/Training: 63188.0
(epoch = 1, round = 7, global round = 7), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 8
(epoch = 1, round = 8, global round = 8), Loss/Training: 1.5671628467738627
(epoch = 1, round = 8, global round = 8), Accuracy/Training: 44.8
reporting (epoch = 1, round = 8, global round = 8) for aggregation


Round:  40%|████      | 8/20 [00:08<00:13,  1.13s/round]

(epoch = 1, round = 8, global round = 8), Loss/Aggregation: 1.5792522147297858
(epoch = 1, round = 8, global round = 8), Accuracy/Aggregation: 42.88
(epoch = 1, round = 8, global round = 8), Client to Server Bytes Sent/Training: 63188.0
(epoch = 1, round = 8, global round = 8), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 9
(epoch = 1, round = 9, global round = 9), Loss/Training: 1.5033099383115769
(epoch = 1, round = 9, global round = 9), Accuracy/Training: 46.6
reporting (epoch = 1, round = 9, global round = 9) for aggregation


Round:  45%|████▌     | 9/20 [00:10<00:12,  1.11s/round]

(epoch = 1, round = 9, global round = 9), Loss/Aggregation: 1.6414472125470638
(epoch = 1, round = 9, global round = 9), Accuracy/Aggregation: 45.0
(epoch = 1, round = 9, global round = 9), Client to Server Bytes Sent/Training: 63188.0
(epoch = 1, round = 9, global round = 9), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 10
(epoch = 1, round = 10, global round = 10), Loss/Training: 1.475800259411335
(epoch = 1, round = 10, global round = 10), Accuracy/Training: 47.4
reporting (epoch = 1, round = 10, global round = 10) for aggregation


Round:  50%|█████     | 10/20 [00:11<00:11,  1.10s/round]

(epoch = 1, round = 10, global round = 10), Loss/Aggregation: 1.4843108616769314
(epoch = 1, round = 10, global round = 10), Accuracy/Aggregation: 46.2
(epoch = 1, round = 10, global round = 10), Client to Server Bytes Sent/Training: 63188.0
(epoch = 1, round = 10, global round = 10), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 11
(epoch = 1, round = 11, global round = 11), Loss/Training: 1.4825674451887607
(epoch = 1, round = 11, global round = 11), Accuracy/Training: 47.04
reporting (epoch = 1, round = 11, global round = 11) for aggregation


Round:  55%|█████▌    | 11/20 [00:12<00:09,  1.10s/round]

(epoch = 1, round = 11, global round = 11), Loss/Aggregation: 1.6685877218842506
(epoch = 1, round = 11, global round = 11), Accuracy/Aggregation: 40.36
(epoch = 1, round = 11, global round = 11), Client to Server Bytes Sent/Training: 63188.0
(epoch = 1, round = 11, global round = 11), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 12
(epoch = 1, round = 12, global round = 12), Loss/Training: 1.4501659333705903
(epoch = 1, round = 12, global round = 12), Accuracy/Training: 48.36
reporting (epoch = 1, round = 12, global round = 12) for aggregation


Round:  60%|██████    | 12/20 [00:13<00:08,  1.10s/round]

(epoch = 1, round = 12, global round = 12), Loss/Aggregation: 1.4886769562959672
(epoch = 1, round = 12, global round = 12), Accuracy/Aggregation: 46.6
(epoch = 1, round = 12, global round = 12), Client to Server Bytes Sent/Training: 63188.0
(epoch = 1, round = 12, global round = 12), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 13
(epoch = 1, round = 13, global round = 13), Loss/Training: 1.3928521186113358
(epoch = 1, round = 13, global round = 13), Accuracy/Training: 50.88
reporting (epoch = 1, round = 13, global round = 13) for aggregation


Round:  65%|██████▌   | 13/20 [00:14<00:07,  1.09s/round]

(epoch = 1, round = 13, global round = 13), Loss/Aggregation: 1.4214412972331048
(epoch = 1, round = 13, global round = 13), Accuracy/Aggregation: 49.84
(epoch = 1, round = 13, global round = 13), Client to Server Bytes Sent/Training: 63188.0
(epoch = 1, round = 13, global round = 13), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 14
(epoch = 1, round = 14, global round = 14), Loss/Training: 1.3448735237121583
(epoch = 1, round = 14, global round = 14), Accuracy/Training: 51.92
reporting (epoch = 1, round = 14, global round = 14) for aggregation


Round:  70%|███████   | 14/20 [00:15<00:06,  1.09s/round]

(epoch = 1, round = 14, global round = 14), Loss/Aggregation: 1.4124967068433762
(epoch = 1, round = 14, global round = 14), Accuracy/Aggregation: 50.6
(epoch = 1, round = 14, global round = 14), Client to Server Bytes Sent/Training: 63188.0
(epoch = 1, round = 14, global round = 14), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 15
(epoch = 1, round = 15, global round = 15), Loss/Training: 1.347032680362463
(epoch = 1, round = 15, global round = 15), Accuracy/Training: 52.12
reporting (epoch = 1, round = 15, global round = 15) for aggregation


Round:  75%|███████▌  | 15/20 [00:16<00:05,  1.09s/round]

(epoch = 1, round = 15, global round = 15), Loss/Aggregation: 1.3111015014350413
(epoch = 1, round = 15, global round = 15), Accuracy/Aggregation: 53.48
(epoch = 1, round = 15, global round = 15), Client to Server Bytes Sent/Training: 63188.0
(epoch = 1, round = 15, global round = 15), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 16
(epoch = 1, round = 16, global round = 16), Loss/Training: 1.2933460563421249
(epoch = 1, round = 16, global round = 16), Accuracy/Training: 53.0
reporting (epoch = 1, round = 16, global round = 16) for aggregation


Round:  80%|████████  | 16/20 [00:17<00:04,  1.09s/round]

(epoch = 1, round = 16, global round = 16), Loss/Aggregation: 1.3642201125621796
(epoch = 1, round = 16, global round = 16), Accuracy/Aggregation: 50.8
(epoch = 1, round = 16, global round = 16), Client to Server Bytes Sent/Training: 63188.0
(epoch = 1, round = 16, global round = 16), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 17
(epoch = 1, round = 17, global round = 17), Loss/Training: 1.3323160901665687
(epoch = 1, round = 17, global round = 17), Accuracy/Training: 52.28
reporting (epoch = 1, round = 17, global round = 17) for aggregation


Round:  85%|████████▌ | 17/20 [00:18<00:03,  1.09s/round]

(epoch = 1, round = 17, global round = 17), Loss/Aggregation: 1.310161143541336
(epoch = 1, round = 17, global round = 17), Accuracy/Aggregation: 54.6
(epoch = 1, round = 17, global round = 17), Client to Server Bytes Sent/Training: 63188.0
(epoch = 1, round = 17, global round = 17), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 18
(epoch = 1, round = 18, global round = 18), Loss/Training: 1.2704973936080932
(epoch = 1, round = 18, global round = 18), Accuracy/Training: 56.0
reporting (epoch = 1, round = 18, global round = 18) for aggregation


Round:  90%|█████████ | 18/20 [00:19<00:02,  1.09s/round]

(epoch = 1, round = 18, global round = 18), Loss/Aggregation: 1.355839768052101
(epoch = 1, round = 18, global round = 18), Accuracy/Aggregation: 51.28
(epoch = 1, round = 18, global round = 18), Client to Server Bytes Sent/Training: 63188.0
(epoch = 1, round = 18, global round = 18), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 19
(epoch = 1, round = 19, global round = 19), Loss/Training: 1.258171808719635
(epoch = 1, round = 19, global round = 19), Accuracy/Training: 55.08
reporting (epoch = 1, round = 19, global round = 19) for aggregation


Round:  95%|█████████▌| 19/20 [00:20<00:01,  1.09s/round]

(epoch = 1, round = 19, global round = 19), Loss/Aggregation: 1.26211207062006
(epoch = 1, round = 19, global round = 19), Accuracy/Aggregation: 55.2
(epoch = 1, round = 19, global round = 19), Client to Server Bytes Sent/Training: 63188.0
(epoch = 1, round = 19, global round = 19), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 20
(epoch = 1, round = 20, global round = 20), Loss/Training: 1.2380948312580586
(epoch = 1, round = 20, global round = 20), Accuracy/Training: 56.96
reporting (epoch = 1, round = 20, global round = 20) for aggregation
(epoch = 1, round = 20, global round = 20), Loss/Aggregation: 1.2786135971546173
(epoch = 1, round = 20, global round = 20), Accuracy/Aggregation: 53.52
(epoch = 1, round = 20, global round = 20), Client to Server Bytes Sent/Training: 63188.0
(epoch = 1, round = 20, global round = 20), Server to Client Bytes Sent/Training: 126376.0
Running (epoch = 1, round = 20, global round = 20) for Eval
(epoch = 1, round = 20, glo

Epoch:   0%|          | 0/1 [00:23<?, ?epoch/s]


We now look at the scalar quantization channel, which quantizes tensors in each client update to between 1 and 8 bits. The channel supports both *per-tensor quantization* and *per-channel quantization*. Let us consider the former which is the simplest case, where each tensor shares the same zero-point and scale parameters. The channel configuration can be now modified to select this channel, with its associated per-tensor quantization flag and number of bits (*n_bits*) parameter.

In [46]:
json_config["trainer"]["channel"] = {
  "report_communication_metrics": True,
  "_base_": "base_scalar_quantization_channel",
  "n_bits": 3,
  "quantize_per_tensor": True
}

run_cifar10_fl_training(json_config)

/usr/local/lib/python3.7/dist-packages/hydra/experimental/initialize.py:36: UserWarning: hydra.experimental.initialize() is no longer experimental. Use hydra.initialize()
  message="hydra.experimental.initialize() is no longer experimental."
/usr/local/lib/python3.7/dist-packages/hydra/experimental/compose.py:19: UserWarning: hydra.experimental.compose() is no longer experimental. Use hydra.compose()
  message="hydra.experimental.compose() is no longer experimental."
Round:   0%|          | 0/20 [00:00<?, ?round/s]

Train finished Global Round: 1
(epoch = 1, round = 1, global round = 1), Loss/Training: 2.3375110030174255
(epoch = 1, round = 1, global round = 1), Accuracy/Training: 14.6
reporting (epoch = 1, round = 1, global round = 1) for aggregation


Round:   5%|▌         | 1/20 [00:01<00:20,  1.10s/round]

(epoch = 1, round = 1, global round = 1), Loss/Aggregation: 2.3128815978765487
(epoch = 1, round = 1, global round = 1), Accuracy/Aggregation: 14.64
(epoch = 1, round = 1, global round = 1), Client to Server Bytes Sent/Training: 12640.0
(epoch = 1, round = 1, global round = 1), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 2
(epoch = 1, round = 2, global round = 2), Loss/Training: 2.1466801568865774
(epoch = 1, round = 2, global round = 2), Accuracy/Training: 21.52
reporting (epoch = 1, round = 2, global round = 2) for aggregation


Round:  10%|█         | 2/20 [00:02<00:19,  1.11s/round]

(epoch = 1, round = 2, global round = 2), Loss/Aggregation: 2.0744873866438867
(epoch = 1, round = 2, global round = 2), Accuracy/Aggregation: 25.68
(epoch = 1, round = 2, global round = 2), Client to Server Bytes Sent/Training: 12640.0
(epoch = 1, round = 2, global round = 2), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 3
(epoch = 1, round = 3, global round = 3), Loss/Training: 1.9881595760583877
(epoch = 1, round = 3, global round = 3), Accuracy/Training: 29.0
reporting (epoch = 1, round = 3, global round = 3) for aggregation


Round:  15%|█▌        | 3/20 [00:03<00:18,  1.11s/round]

(epoch = 1, round = 3, global round = 3), Loss/Aggregation: 1.8731757253408432
(epoch = 1, round = 3, global round = 3), Accuracy/Aggregation: 31.72
(epoch = 1, round = 3, global round = 3), Client to Server Bytes Sent/Training: 12640.0
(epoch = 1, round = 3, global round = 3), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 4
(epoch = 1, round = 4, global round = 4), Loss/Training: 1.8633052006363868
(epoch = 1, round = 4, global round = 4), Accuracy/Training: 35.0
reporting (epoch = 1, round = 4, global round = 4) for aggregation


Round:  20%|██        | 4/20 [00:04<00:17,  1.11s/round]

(epoch = 1, round = 4, global round = 4), Loss/Aggregation: 1.7945417031645774
(epoch = 1, round = 4, global round = 4), Accuracy/Aggregation: 34.48
(epoch = 1, round = 4, global round = 4), Client to Server Bytes Sent/Training: 12640.0
(epoch = 1, round = 4, global round = 4), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 5
(epoch = 1, round = 5, global round = 5), Loss/Training: 1.7407176464796066
(epoch = 1, round = 5, global round = 5), Accuracy/Training: 36.76
reporting (epoch = 1, round = 5, global round = 5) for aggregation


Round:  25%|██▌       | 5/20 [00:05<00:16,  1.10s/round]

(epoch = 1, round = 5, global round = 5), Loss/Aggregation: 1.755663387477398
(epoch = 1, round = 5, global round = 5), Accuracy/Aggregation: 35.12
(epoch = 1, round = 5, global round = 5), Client to Server Bytes Sent/Training: 12640.0
(epoch = 1, round = 5, global round = 5), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 6
(epoch = 1, round = 6, global round = 6), Loss/Training: 1.6583320304751397
(epoch = 1, round = 6, global round = 6), Accuracy/Training: 40.28
reporting (epoch = 1, round = 6, global round = 6) for aggregation


Round:  30%|███       | 6/20 [00:06<00:15,  1.10s/round]

(epoch = 1, round = 6, global round = 6), Loss/Aggregation: 1.5651693150401116
(epoch = 1, round = 6, global round = 6), Accuracy/Aggregation: 44.0
(epoch = 1, round = 6, global round = 6), Client to Server Bytes Sent/Training: 12640.0
(epoch = 1, round = 6, global round = 6), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 7
(epoch = 1, round = 7, global round = 7), Loss/Training: 1.5631324753165246
(epoch = 1, round = 7, global round = 7), Accuracy/Training: 43.72
reporting (epoch = 1, round = 7, global round = 7) for aggregation


Round:  35%|███▌      | 7/20 [00:07<00:14,  1.11s/round]

(epoch = 1, round = 7, global round = 7), Loss/Aggregation: 1.6097149014472962
(epoch = 1, round = 7, global round = 7), Accuracy/Aggregation: 43.04
(epoch = 1, round = 7, global round = 7), Client to Server Bytes Sent/Training: 12640.0
(epoch = 1, round = 7, global round = 7), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 8
(epoch = 1, round = 8, global round = 8), Loss/Training: 1.6020618483424187
(epoch = 1, round = 8, global round = 8), Accuracy/Training: 43.56
reporting (epoch = 1, round = 8, global round = 8) for aggregation


Round:  40%|████      | 8/20 [00:08<00:13,  1.11s/round]

(epoch = 1, round = 8, global round = 8), Loss/Aggregation: 1.577898009121418
(epoch = 1, round = 8, global round = 8), Accuracy/Aggregation: 44.24
(epoch = 1, round = 8, global round = 8), Client to Server Bytes Sent/Training: 12640.0
(epoch = 1, round = 8, global round = 8), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 9
(epoch = 1, round = 9, global round = 9), Loss/Training: 1.5317140854895115
(epoch = 1, round = 9, global round = 9), Accuracy/Training: 45.32
reporting (epoch = 1, round = 9, global round = 9) for aggregation


Round:  45%|████▌     | 9/20 [00:09<00:12,  1.11s/round]

(epoch = 1, round = 9, global round = 9), Loss/Aggregation: 1.502110907435417
(epoch = 1, round = 9, global round = 9), Accuracy/Aggregation: 46.56
(epoch = 1, round = 9, global round = 9), Client to Server Bytes Sent/Training: 12640.0
(epoch = 1, round = 9, global round = 9), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 10
(epoch = 1, round = 10, global round = 10), Loss/Training: 1.518892516195774
(epoch = 1, round = 10, global round = 10), Accuracy/Training: 45.96
reporting (epoch = 1, round = 10, global round = 10) for aggregation


Round:  50%|█████     | 10/20 [00:11<00:11,  1.11s/round]

(epoch = 1, round = 10, global round = 10), Loss/Aggregation: 1.5109284773468972
(epoch = 1, round = 10, global round = 10), Accuracy/Aggregation: 44.68
(epoch = 1, round = 10, global round = 10), Client to Server Bytes Sent/Training: 12640.0
(epoch = 1, round = 10, global round = 10), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 11
(epoch = 1, round = 11, global round = 11), Loss/Training: 1.4576153606176376
(epoch = 1, round = 11, global round = 11), Accuracy/Training: 47.48
reporting (epoch = 1, round = 11, global round = 11) for aggregation


Round:  55%|█████▌    | 11/20 [00:12<00:09,  1.10s/round]

(epoch = 1, round = 11, global round = 11), Loss/Aggregation: 1.4031564801931382
(epoch = 1, round = 11, global round = 11), Accuracy/Aggregation: 49.08
(epoch = 1, round = 11, global round = 11), Client to Server Bytes Sent/Training: 12640.0
(epoch = 1, round = 11, global round = 11), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 12
(epoch = 1, round = 12, global round = 12), Loss/Training: 1.4265179440379143
(epoch = 1, round = 12, global round = 12), Accuracy/Training: 47.96
reporting (epoch = 1, round = 12, global round = 12) for aggregation


Round:  60%|██████    | 12/20 [00:13<00:08,  1.11s/round]

(epoch = 1, round = 12, global round = 12), Loss/Aggregation: 1.3959756180644036
(epoch = 1, round = 12, global round = 12), Accuracy/Aggregation: 49.2
(epoch = 1, round = 12, global round = 12), Client to Server Bytes Sent/Training: 12640.0
(epoch = 1, round = 12, global round = 12), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 13
(epoch = 1, round = 13, global round = 13), Loss/Training: 1.4125475771725178
(epoch = 1, round = 13, global round = 13), Accuracy/Training: 50.12
reporting (epoch = 1, round = 13, global round = 13) for aggregation


Round:  65%|██████▌   | 13/20 [00:14<00:07,  1.11s/round]

(epoch = 1, round = 13, global round = 13), Loss/Aggregation: 1.346294953674078
(epoch = 1, round = 13, global round = 13), Accuracy/Aggregation: 52.96
(epoch = 1, round = 13, global round = 13), Client to Server Bytes Sent/Training: 12640.0
(epoch = 1, round = 13, global round = 13), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 14
(epoch = 1, round = 14, global round = 14), Loss/Training: 1.3793655537068843
(epoch = 1, round = 14, global round = 14), Accuracy/Training: 50.32
reporting (epoch = 1, round = 14, global round = 14) for aggregation


Round:  70%|███████   | 14/20 [00:15<00:06,  1.11s/round]

(epoch = 1, round = 14, global round = 14), Loss/Aggregation: 1.402449794858694
(epoch = 1, round = 14, global round = 14), Accuracy/Aggregation: 49.24
(epoch = 1, round = 14, global round = 14), Client to Server Bytes Sent/Training: 12640.0
(epoch = 1, round = 14, global round = 14), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 15
(epoch = 1, round = 15, global round = 15), Loss/Training: 1.3876709729433059
(epoch = 1, round = 15, global round = 15), Accuracy/Training: 49.64
reporting (epoch = 1, round = 15, global round = 15) for aggregation


Round:  75%|███████▌  | 15/20 [00:16<00:05,  1.11s/round]

(epoch = 1, round = 15, global round = 15), Loss/Aggregation: 1.3408145748078824
(epoch = 1, round = 15, global round = 15), Accuracy/Aggregation: 51.44
(epoch = 1, round = 15, global round = 15), Client to Server Bytes Sent/Training: 12640.0
(epoch = 1, round = 15, global round = 15), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 16
(epoch = 1, round = 16, global round = 16), Loss/Training: 1.3274762369692326
(epoch = 1, round = 16, global round = 16), Accuracy/Training: 52.2
reporting (epoch = 1, round = 16, global round = 16) for aggregation


Round:  80%|████████  | 16/20 [00:17<00:04,  1.11s/round]

(epoch = 1, round = 16, global round = 16), Loss/Aggregation: 1.3070026203989982
(epoch = 1, round = 16, global round = 16), Accuracy/Aggregation: 53.36
(epoch = 1, round = 16, global round = 16), Client to Server Bytes Sent/Training: 12640.0
(epoch = 1, round = 16, global round = 16), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 17
(epoch = 1, round = 17, global round = 17), Loss/Training: 1.3044292904436587
(epoch = 1, round = 17, global round = 17), Accuracy/Training: 52.6
reporting (epoch = 1, round = 17, global round = 17) for aggregation


Round:  85%|████████▌ | 17/20 [00:18<00:03,  1.11s/round]

(epoch = 1, round = 17, global round = 17), Loss/Aggregation: 1.3029765531420707
(epoch = 1, round = 17, global round = 17), Accuracy/Aggregation: 53.16
(epoch = 1, round = 17, global round = 17), Client to Server Bytes Sent/Training: 12640.0
(epoch = 1, round = 17, global round = 17), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 18
(epoch = 1, round = 18, global round = 18), Loss/Training: 1.2869433745741845
(epoch = 1, round = 18, global round = 18), Accuracy/Training: 53.76
reporting (epoch = 1, round = 18, global round = 18) for aggregation


Round:  90%|█████████ | 18/20 [00:19<00:02,  1.11s/round]

(epoch = 1, round = 18, global round = 18), Loss/Aggregation: 1.2921988151967525
(epoch = 1, round = 18, global round = 18), Accuracy/Aggregation: 53.04
(epoch = 1, round = 18, global round = 18), Client to Server Bytes Sent/Training: 12640.0
(epoch = 1, round = 18, global round = 18), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 19
(epoch = 1, round = 19, global round = 19), Loss/Training: 1.3031659096479415
(epoch = 1, round = 19, global round = 19), Accuracy/Training: 54.52
reporting (epoch = 1, round = 19, global round = 19) for aggregation


Round:  95%|█████████▌| 19/20 [00:21<00:01,  1.11s/round]

(epoch = 1, round = 19, global round = 19), Loss/Aggregation: 1.349413202702999
(epoch = 1, round = 19, global round = 19), Accuracy/Aggregation: 52.04
(epoch = 1, round = 19, global round = 19), Client to Server Bytes Sent/Training: 12640.0
(epoch = 1, round = 19, global round = 19), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 20
(epoch = 1, round = 20, global round = 20), Loss/Training: 1.2935495600104332
(epoch = 1, round = 20, global round = 20), Accuracy/Training: 53.08
reporting (epoch = 1, round = 20, global round = 20) for aggregation
(epoch = 1, round = 20, global round = 20), Loss/Aggregation: 1.3623090006411076
(epoch = 1, round = 20, global round = 20), Accuracy/Aggregation: 52.32
(epoch = 1, round = 20, global round = 20), Client to Server Bytes Sent/Training: 12640.0
(epoch = 1, round = 20, global round = 20), Server to Client Bytes Sent/Training: 126376.0
Running (epoch = 1, round = 20, global round = 20) for Eval
(epoch = 1, round = 20, g

Epoch:   0%|          | 0/1 [00:23<?, ?epoch/s]


Note how the number of bytes during upload is not exactly 11847 (=126376 * 3/32). The additional overhead of 793 bytes is due to the upload messages also including scale and zero-point information, which is necessary server-side to dequantize the client update tensors. We now wrap up this tutorial by looking at another channel, [sparse mask channel](https://github.com/facebookresearch/FLSim/blob/main/channels/sparse_mask_channel.py) which implements sparsity on the client updates. The sparsity mask is client-specific, and the simulator supports two approaches to determine this mask : *random* and *Top-K*. Random mask is self-explanatory, with the per-tensor sparsity mask being set randomly based on the sparsity level, and *Top-K* sparsity masking out the lowest magnitude elements within each model update tensor. Let us now define a config using the Top-K mask with a sparsity level of 99%, and run a round of training using this compression. 

In [47]:
json_config["trainer"]["channel"] = {
  "report_communication_metrics": True,
  "_base_": "base_sparse_mask_channel",
  "proportion_of_zero_weights":0.99,
  "sparsity_method": "topk"
}

run_cifar10_fl_training(json_config)

/usr/local/lib/python3.7/dist-packages/hydra/experimental/initialize.py:36: UserWarning: hydra.experimental.initialize() is no longer experimental. Use hydra.initialize()
  message="hydra.experimental.initialize() is no longer experimental."
/usr/local/lib/python3.7/dist-packages/hydra/experimental/compose.py:19: UserWarning: hydra.experimental.compose() is no longer experimental. Use hydra.compose()
  message="hydra.experimental.compose() is no longer experimental."
Round:   0%|          | 0/20 [00:00<?, ?round/s]

Train finished Global Round: 1
(epoch = 1, round = 1, global round = 1), Loss/Training: 2.3019975513219832
(epoch = 1, round = 1, global round = 1), Accuracy/Training: 13.84
reporting (epoch = 1, round = 1, global round = 1) for aggregation


Round:   5%|▌         | 1/20 [00:01<00:21,  1.11s/round]

(epoch = 1, round = 1, global round = 1), Loss/Aggregation: 2.362009060382843
(epoch = 1, round = 1, global round = 1), Accuracy/Aggregation: 12.12
(epoch = 1, round = 1, global round = 1), Client to Server Bytes Sent/Training: 5245.25
(epoch = 1, round = 1, global round = 1), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 2
(epoch = 1, round = 2, global round = 2), Loss/Training: 2.2777017533779143
(epoch = 1, round = 2, global round = 2), Accuracy/Training: 14.6
reporting (epoch = 1, round = 2, global round = 2) for aggregation


Round:  10%|█         | 2/20 [00:02<00:20,  1.12s/round]

(epoch = 1, round = 2, global round = 2), Loss/Aggregation: 2.224752974510193
(epoch = 1, round = 2, global round = 2), Accuracy/Aggregation: 17.8
(epoch = 1, round = 2, global round = 2), Client to Server Bytes Sent/Training: 5245.25
(epoch = 1, round = 2, global round = 2), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 3
(epoch = 1, round = 3, global round = 3), Loss/Training: 2.2061293333768845
(epoch = 1, round = 3, global round = 3), Accuracy/Training: 18.56
reporting (epoch = 1, round = 3, global round = 3) for aggregation


Round:  15%|█▌        | 3/20 [00:03<00:19,  1.12s/round]

(epoch = 1, round = 3, global round = 3), Loss/Aggregation: 2.148963275551796
(epoch = 1, round = 3, global round = 3), Accuracy/Aggregation: 22.08
(epoch = 1, round = 3, global round = 3), Client to Server Bytes Sent/Training: 5245.25
(epoch = 1, round = 3, global round = 3), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 4
(epoch = 1, round = 4, global round = 4), Loss/Training: 2.1595217913389204
(epoch = 1, round = 4, global round = 4), Accuracy/Training: 19.44
reporting (epoch = 1, round = 4, global round = 4) for aggregation


Round:  20%|██        | 4/20 [00:04<00:17,  1.11s/round]

(epoch = 1, round = 4, global round = 4), Loss/Aggregation: 2.1399784609675407
(epoch = 1, round = 4, global round = 4), Accuracy/Aggregation: 18.6
(epoch = 1, round = 4, global round = 4), Client to Server Bytes Sent/Training: 5245.25
(epoch = 1, round = 4, global round = 4), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 5
(epoch = 1, round = 5, global round = 5), Loss/Training: 2.086236535012722
(epoch = 1, round = 5, global round = 5), Accuracy/Training: 23.88
reporting (epoch = 1, round = 5, global round = 5) for aggregation


Round:  25%|██▌       | 5/20 [00:05<00:16,  1.12s/round]

(epoch = 1, round = 5, global round = 5), Loss/Aggregation: 2.0658847823739053
(epoch = 1, round = 5, global round = 5), Accuracy/Aggregation: 23.08
(epoch = 1, round = 5, global round = 5), Client to Server Bytes Sent/Training: 5245.25
(epoch = 1, round = 5, global round = 5), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 6
(epoch = 1, round = 6, global round = 6), Loss/Training: 2.0493230774998663
(epoch = 1, round = 6, global round = 6), Accuracy/Training: 24.84
reporting (epoch = 1, round = 6, global round = 6) for aggregation


Round:  30%|███       | 6/20 [00:06<00:15,  1.11s/round]

(epoch = 1, round = 6, global round = 6), Loss/Aggregation: 2.0185644060373304
(epoch = 1, round = 6, global round = 6), Accuracy/Aggregation: 26.4
(epoch = 1, round = 6, global round = 6), Client to Server Bytes Sent/Training: 5245.25
(epoch = 1, round = 6, global round = 6), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 7
(epoch = 1, round = 7, global round = 7), Loss/Training: 1.9813435301184654
(epoch = 1, round = 7, global round = 7), Accuracy/Training: 27.64
reporting (epoch = 1, round = 7, global round = 7) for aggregation


Round:  35%|███▌      | 7/20 [00:07<00:14,  1.12s/round]

(epoch = 1, round = 7, global round = 7), Loss/Aggregation: 1.935346533358097
(epoch = 1, round = 7, global round = 7), Accuracy/Aggregation: 29.24
(epoch = 1, round = 7, global round = 7), Client to Server Bytes Sent/Training: 5245.25
(epoch = 1, round = 7, global round = 7), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 8
(epoch = 1, round = 8, global round = 8), Loss/Training: 1.9402765318751336
(epoch = 1, round = 8, global round = 8), Accuracy/Training: 29.72
reporting (epoch = 1, round = 8, global round = 8) for aggregation


Round:  40%|████      | 8/20 [00:08<00:13,  1.12s/round]

(epoch = 1, round = 8, global round = 8), Loss/Aggregation: 1.8961942180991174
(epoch = 1, round = 8, global round = 8), Accuracy/Aggregation: 30.88
(epoch = 1, round = 8, global round = 8), Client to Server Bytes Sent/Training: 5245.25
(epoch = 1, round = 8, global round = 8), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 9
(epoch = 1, round = 9, global round = 9), Loss/Training: 1.8993749469518661
(epoch = 1, round = 9, global round = 9), Accuracy/Training: 31.04
reporting (epoch = 1, round = 9, global round = 9) for aggregation


Round:  45%|████▌     | 9/20 [00:10<00:12,  1.12s/round]

(epoch = 1, round = 9, global round = 9), Loss/Aggregation: 1.8608284696936608
(epoch = 1, round = 9, global round = 9), Accuracy/Aggregation: 33.04
(epoch = 1, round = 9, global round = 9), Client to Server Bytes Sent/Training: 5245.25
(epoch = 1, round = 9, global round = 9), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 10
(epoch = 1, round = 10, global round = 10), Loss/Training: 1.8336709573864938
(epoch = 1, round = 10, global round = 10), Accuracy/Training: 34.64
reporting (epoch = 1, round = 10, global round = 10) for aggregation


Round:  50%|█████     | 10/20 [00:11<00:11,  1.12s/round]

(epoch = 1, round = 10, global round = 10), Loss/Aggregation: 1.8024304702877998
(epoch = 1, round = 10, global round = 10), Accuracy/Aggregation: 35.44
(epoch = 1, round = 10, global round = 10), Client to Server Bytes Sent/Training: 5245.25
(epoch = 1, round = 10, global round = 10), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 11
(epoch = 1, round = 11, global round = 11), Loss/Training: 1.8321253091096878
(epoch = 1, round = 11, global round = 11), Accuracy/Training: 36.36
reporting (epoch = 1, round = 11, global round = 11) for aggregation


Round:  55%|█████▌    | 11/20 [00:12<00:10,  1.12s/round]

(epoch = 1, round = 11, global round = 11), Loss/Aggregation: 1.7663378909230232
(epoch = 1, round = 11, global round = 11), Accuracy/Aggregation: 37.64
(epoch = 1, round = 11, global round = 11), Client to Server Bytes Sent/Training: 5245.25
(epoch = 1, round = 11, global round = 11), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 12
(epoch = 1, round = 12, global round = 12), Loss/Training: 1.801997783780098
(epoch = 1, round = 12, global round = 12), Accuracy/Training: 35.56
reporting (epoch = 1, round = 12, global round = 12) for aggregation


Round:  60%|██████    | 12/20 [00:13<00:08,  1.11s/round]

(epoch = 1, round = 12, global round = 12), Loss/Aggregation: 1.745076698064804
(epoch = 1, round = 12, global round = 12), Accuracy/Aggregation: 38.92
(epoch = 1, round = 12, global round = 12), Client to Server Bytes Sent/Training: 5245.25
(epoch = 1, round = 12, global round = 12), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 13
(epoch = 1, round = 13, global round = 13), Loss/Training: 1.7723519757390023
(epoch = 1, round = 13, global round = 13), Accuracy/Training: 36.16
reporting (epoch = 1, round = 13, global round = 13) for aggregation


Round:  65%|██████▌   | 13/20 [00:14<00:07,  1.12s/round]

(epoch = 1, round = 13, global round = 13), Loss/Aggregation: 1.7055876404047012
(epoch = 1, round = 13, global round = 13), Accuracy/Aggregation: 39.8
(epoch = 1, round = 13, global round = 13), Client to Server Bytes Sent/Training: 5245.25
(epoch = 1, round = 13, global round = 13), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 14
(epoch = 1, round = 14, global round = 14), Loss/Training: 1.778145718574524
(epoch = 1, round = 14, global round = 14), Accuracy/Training: 36.12
reporting (epoch = 1, round = 14, global round = 14) for aggregation


Round:  70%|███████   | 14/20 [00:15<00:06,  1.11s/round]

(epoch = 1, round = 14, global round = 14), Loss/Aggregation: 1.7025674641132356
(epoch = 1, round = 14, global round = 14), Accuracy/Aggregation: 39.56
(epoch = 1, round = 14, global round = 14), Client to Server Bytes Sent/Training: 5245.25
(epoch = 1, round = 14, global round = 14), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 15
(epoch = 1, round = 15, global round = 15), Loss/Training: 1.7285629630088806
(epoch = 1, round = 15, global round = 15), Accuracy/Training: 37.76
reporting (epoch = 1, round = 15, global round = 15) for aggregation


Round:  75%|███████▌  | 15/20 [00:16<00:05,  1.12s/round]

(epoch = 1, round = 15, global round = 15), Loss/Aggregation: 1.6619032606482507
(epoch = 1, round = 15, global round = 15), Accuracy/Aggregation: 39.56
(epoch = 1, round = 15, global round = 15), Client to Server Bytes Sent/Training: 5245.25
(epoch = 1, round = 15, global round = 15), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 16
(epoch = 1, round = 16, global round = 16), Loss/Training: 1.7130748063325882
(epoch = 1, round = 16, global round = 16), Accuracy/Training: 39.24
reporting (epoch = 1, round = 16, global round = 16) for aggregation


Round:  80%|████████  | 16/20 [00:17<00:04,  1.11s/round]

(epoch = 1, round = 16, global round = 16), Loss/Aggregation: 1.646116641163826
(epoch = 1, round = 16, global round = 16), Accuracy/Aggregation: 40.16
(epoch = 1, round = 16, global round = 16), Client to Server Bytes Sent/Training: 5245.25
(epoch = 1, round = 16, global round = 16), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 17
(epoch = 1, round = 17, global round = 17), Loss/Training: 1.6662024706602097
(epoch = 1, round = 17, global round = 17), Accuracy/Training: 40.08
reporting (epoch = 1, round = 17, global round = 17) for aggregation


Round:  85%|████████▌ | 17/20 [00:18<00:03,  1.11s/round]

(epoch = 1, round = 17, global round = 17), Loss/Aggregation: 1.5840383544564247
(epoch = 1, round = 17, global round = 17), Accuracy/Aggregation: 42.76
(epoch = 1, round = 17, global round = 17), Client to Server Bytes Sent/Training: 5245.25
(epoch = 1, round = 17, global round = 17), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 18
(epoch = 1, round = 18, global round = 18), Loss/Training: 1.6836617529392242
(epoch = 1, round = 18, global round = 18), Accuracy/Training: 40.52
reporting (epoch = 1, round = 18, global round = 18) for aggregation


Round:  90%|█████████ | 18/20 [00:20<00:02,  1.12s/round]

(epoch = 1, round = 18, global round = 18), Loss/Aggregation: 1.630456741154194
(epoch = 1, round = 18, global round = 18), Accuracy/Aggregation: 40.52
(epoch = 1, round = 18, global round = 18), Client to Server Bytes Sent/Training: 5245.25
(epoch = 1, round = 18, global round = 18), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 19
(epoch = 1, round = 19, global round = 19), Loss/Training: 1.6380363792181014
(epoch = 1, round = 19, global round = 19), Accuracy/Training: 41.08
reporting (epoch = 1, round = 19, global round = 19) for aggregation


Round:  95%|█████████▌| 19/20 [00:21<00:01,  1.11s/round]

(epoch = 1, round = 19, global round = 19), Loss/Aggregation: 1.5991919979453086
(epoch = 1, round = 19, global round = 19), Accuracy/Aggregation: 42.72
(epoch = 1, round = 19, global round = 19), Client to Server Bytes Sent/Training: 5245.25
(epoch = 1, round = 19, global round = 19), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 20
(epoch = 1, round = 20, global round = 20), Loss/Training: 1.6388011947274208
(epoch = 1, round = 20, global round = 20), Accuracy/Training: 40.4
reporting (epoch = 1, round = 20, global round = 20) for aggregation
(epoch = 1, round = 20, global round = 20), Loss/Aggregation: 1.5713405102491378
(epoch = 1, round = 20, global round = 20), Accuracy/Aggregation: 43.64
(epoch = 1, round = 20, global round = 20), Client to Server Bytes Sent/Training: 5245.25
(epoch = 1, round = 20, global round = 20), Server to Client Bytes Sent/Training: 126376.0
Running (epoch = 1, round = 20, global round = 20) for Eval
(epoch = 1, round = 20, g

Epoch:   0%|          | 0/1 [00:23<?, ?epoch/s]


While the non-sparse parameters consume only 1263 bytes, we also need to pack in additional information about the sparsity pattern per tensor. The simulator supports two ways to do this: *bitmask* and *coo*. In the bitmask setting, the sparsity pattern is transported as a concatenated bitstream. We also support another mode *coo*, which follows the popular COO format for encoding sparse tensors as described in [torch.sparse](https://pytorch.org/docs/stable/sparse.html#sparse-coo-tensors) documentation. Here, the non-sparse tensor weights are transmitted along with their 64-bit integer coordinates. Generally *bitmask* has a lower overhead than *coo* in terms of bytes transmitted. Let us conclude our tutorial by running a round of Cifar-10 training with *coo* bytes measurement.

In [48]:
json_config["trainer"]["channel"] = {
  "report_communication_metrics": True,
  "_base_": "base_sparse_mask_channel",
  "proportion_of_zero_weights": 0.99,
  "sparsity_method": "topk",
  "compressed_size_measurement": "coo"
}

run_cifar10_fl_training(json_config)

/usr/local/lib/python3.7/dist-packages/hydra/experimental/initialize.py:36: UserWarning: hydra.experimental.initialize() is no longer experimental. Use hydra.initialize()
  message="hydra.experimental.initialize() is no longer experimental."
/usr/local/lib/python3.7/dist-packages/hydra/experimental/compose.py:19: UserWarning: hydra.experimental.compose() is no longer experimental. Use hydra.compose()
  message="hydra.experimental.compose() is no longer experimental."
Round:   0%|          | 0/20 [00:00<?, ?round/s]

Train finished Global Round: 1
(epoch = 1, round = 1, global round = 1), Loss/Training: 2.301069360971451
(epoch = 1, round = 1, global round = 1), Accuracy/Training: 14.72
reporting (epoch = 1, round = 1, global round = 1) for aggregation


Round:   5%|▌         | 1/20 [00:01<00:21,  1.14s/round]

(epoch = 1, round = 1, global round = 1), Loss/Aggregation: 2.3707367897033693
(epoch = 1, round = 1, global round = 1), Accuracy/Aggregation: 14.04
(epoch = 1, round = 1, global round = 1), Client to Server Bytes Sent/Training: 11032.0
(epoch = 1, round = 1, global round = 1), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 2
(epoch = 1, round = 2, global round = 2), Loss/Training: 2.2785249024629595
(epoch = 1, round = 2, global round = 2), Accuracy/Training: 15.04
reporting (epoch = 1, round = 2, global round = 2) for aggregation


Round:  10%|█         | 2/20 [00:02<00:20,  1.13s/round]

(epoch = 1, round = 2, global round = 2), Loss/Aggregation: 2.2305319935083388
(epoch = 1, round = 2, global round = 2), Accuracy/Aggregation: 16.96
(epoch = 1, round = 2, global round = 2), Client to Server Bytes Sent/Training: 11032.0
(epoch = 1, round = 2, global round = 2), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 3
(epoch = 1, round = 3, global round = 3), Loss/Training: 2.20278856754303
(epoch = 1, round = 3, global round = 3), Accuracy/Training: 18.44
reporting (epoch = 1, round = 3, global round = 3) for aggregation


Round:  15%|█▌        | 3/20 [00:03<00:19,  1.12s/round]

(epoch = 1, round = 3, global round = 3), Loss/Aggregation: 2.1114589139819144
(epoch = 1, round = 3, global round = 3), Accuracy/Aggregation: 25.08
(epoch = 1, round = 3, global round = 3), Client to Server Bytes Sent/Training: 11032.0
(epoch = 1, round = 3, global round = 3), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 4
(epoch = 1, round = 4, global round = 4), Loss/Training: 2.0972865894436836
(epoch = 1, round = 4, global round = 4), Accuracy/Training: 24.24
reporting (epoch = 1, round = 4, global round = 4) for aggregation


Round:  20%|██        | 4/20 [00:04<00:17,  1.12s/round]

(epoch = 1, round = 4, global round = 4), Loss/Aggregation: 2.023341615498066
(epoch = 1, round = 4, global round = 4), Accuracy/Aggregation: 25.64
(epoch = 1, round = 4, global round = 4), Client to Server Bytes Sent/Training: 11032.0
(epoch = 1, round = 4, global round = 4), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 5
(epoch = 1, round = 5, global round = 5), Loss/Training: 2.0499159902334214
(epoch = 1, round = 5, global round = 5), Accuracy/Training: 25.48
reporting (epoch = 1, round = 5, global round = 5) for aggregation


Round:  25%|██▌       | 5/20 [00:05<00:16,  1.12s/round]

(epoch = 1, round = 5, global round = 5), Loss/Aggregation: 2.0058013558387757
(epoch = 1, round = 5, global round = 5), Accuracy/Aggregation: 26.76
(epoch = 1, round = 5, global round = 5), Client to Server Bytes Sent/Training: 11032.0
(epoch = 1, round = 5, global round = 5), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 6
(epoch = 1, round = 6, global round = 6), Loss/Training: 2.020043097436428
(epoch = 1, round = 6, global round = 6), Accuracy/Training: 26.8
reporting (epoch = 1, round = 6, global round = 6) for aggregation


Round:  30%|███       | 6/20 [00:06<00:15,  1.12s/round]

(epoch = 1, round = 6, global round = 6), Loss/Aggregation: 1.9781676337122918
(epoch = 1, round = 6, global round = 6), Accuracy/Aggregation: 28.04
(epoch = 1, round = 6, global round = 6), Client to Server Bytes Sent/Training: 11032.0
(epoch = 1, round = 6, global round = 6), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 7
(epoch = 1, round = 7, global round = 7), Loss/Training: 1.9786505579948426
(epoch = 1, round = 7, global round = 7), Accuracy/Training: 28.44
reporting (epoch = 1, round = 7, global round = 7) for aggregation


Round:  35%|███▌      | 7/20 [00:07<00:14,  1.11s/round]

(epoch = 1, round = 7, global round = 7), Loss/Aggregation: 1.9499336570501327
(epoch = 1, round = 7, global round = 7), Accuracy/Aggregation: 28.4
(epoch = 1, round = 7, global round = 7), Client to Server Bytes Sent/Training: 11032.0
(epoch = 1, round = 7, global round = 7), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 8
(epoch = 1, round = 8, global round = 8), Loss/Training: 1.933263297379017
(epoch = 1, round = 8, global round = 8), Accuracy/Training: 29.8
reporting (epoch = 1, round = 8, global round = 8) for aggregation


Round:  40%|████      | 8/20 [00:08<00:13,  1.12s/round]

(epoch = 1, round = 8, global round = 8), Loss/Aggregation: 1.8777749046683312
(epoch = 1, round = 8, global round = 8), Accuracy/Aggregation: 31.84
(epoch = 1, round = 8, global round = 8), Client to Server Bytes Sent/Training: 11032.0
(epoch = 1, round = 8, global round = 8), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 9
(epoch = 1, round = 9, global round = 9), Loss/Training: 1.8675140663981438
(epoch = 1, round = 9, global round = 9), Accuracy/Training: 33.0
reporting (epoch = 1, round = 9, global round = 9) for aggregation


Round:  45%|████▌     | 9/20 [00:10<00:12,  1.12s/round]

(epoch = 1, round = 9, global round = 9), Loss/Aggregation: 1.8159892335534096
(epoch = 1, round = 9, global round = 9), Accuracy/Aggregation: 35.12
(epoch = 1, round = 9, global round = 9), Client to Server Bytes Sent/Training: 11032.0
(epoch = 1, round = 9, global round = 9), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 10
(epoch = 1, round = 10, global round = 10), Loss/Training: 1.8636312276124953
(epoch = 1, round = 10, global round = 10), Accuracy/Training: 32.44
reporting (epoch = 1, round = 10, global round = 10) for aggregation


Round:  50%|█████     | 10/20 [00:11<00:11,  1.12s/round]

(epoch = 1, round = 10, global round = 10), Loss/Aggregation: 1.8064094245433808
(epoch = 1, round = 10, global round = 10), Accuracy/Aggregation: 35.12
(epoch = 1, round = 10, global round = 10), Client to Server Bytes Sent/Training: 11032.0
(epoch = 1, round = 10, global round = 10), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 11
(epoch = 1, round = 11, global round = 11), Loss/Training: 1.819228184223175
(epoch = 1, round = 11, global round = 11), Accuracy/Training: 34.4
reporting (epoch = 1, round = 11, global round = 11) for aggregation


Round:  55%|█████▌    | 11/20 [00:12<00:10,  1.12s/round]

(epoch = 1, round = 11, global round = 11), Loss/Aggregation: 1.7655522301793098
(epoch = 1, round = 11, global round = 11), Accuracy/Aggregation: 36.52
(epoch = 1, round = 11, global round = 11), Client to Server Bytes Sent/Training: 11032.0
(epoch = 1, round = 11, global round = 11), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 12
(epoch = 1, round = 12, global round = 12), Loss/Training: 1.7525650903582572
(epoch = 1, round = 12, global round = 12), Accuracy/Training: 36.68
reporting (epoch = 1, round = 12, global round = 12) for aggregation


Round:  60%|██████    | 12/20 [00:13<00:08,  1.12s/round]

(epoch = 1, round = 12, global round = 12), Loss/Aggregation: 1.7014419630169868
(epoch = 1, round = 12, global round = 12), Accuracy/Aggregation: 39.16
(epoch = 1, round = 12, global round = 12), Client to Server Bytes Sent/Training: 11032.0
(epoch = 1, round = 12, global round = 12), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 13
(epoch = 1, round = 13, global round = 13), Loss/Training: 1.733429728448391
(epoch = 1, round = 13, global round = 13), Accuracy/Training: 37.88
reporting (epoch = 1, round = 13, global round = 13) for aggregation


Round:  65%|██████▌   | 13/20 [00:14<00:07,  1.12s/round]

(epoch = 1, round = 13, global round = 13), Loss/Aggregation: 1.646027809381485
(epoch = 1, round = 13, global round = 13), Accuracy/Aggregation: 41.2
(epoch = 1, round = 13, global round = 13), Client to Server Bytes Sent/Training: 11032.0
(epoch = 1, round = 13, global round = 13), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 14
(epoch = 1, round = 14, global round = 14), Loss/Training: 1.7055915609002112
(epoch = 1, round = 14, global round = 14), Accuracy/Training: 37.4
reporting (epoch = 1, round = 14, global round = 14) for aggregation


Round:  70%|███████   | 14/20 [00:15<00:06,  1.16s/round]

(epoch = 1, round = 14, global round = 14), Loss/Aggregation: 1.6430437982082366
(epoch = 1, round = 14, global round = 14), Accuracy/Aggregation: 39.64
(epoch = 1, round = 14, global round = 14), Client to Server Bytes Sent/Training: 11032.0
(epoch = 1, round = 14, global round = 14), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 15
(epoch = 1, round = 15, global round = 15), Loss/Training: 1.6904700875282288
(epoch = 1, round = 15, global round = 15), Accuracy/Training: 38.68
reporting (epoch = 1, round = 15, global round = 15) for aggregation


Round:  75%|███████▌  | 15/20 [00:17<00:06,  1.22s/round]

(epoch = 1, round = 15, global round = 15), Loss/Aggregation: 1.6237804561853408
(epoch = 1, round = 15, global round = 15), Accuracy/Aggregation: 41.6
(epoch = 1, round = 15, global round = 15), Client to Server Bytes Sent/Training: 11032.0
(epoch = 1, round = 15, global round = 15), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 16
(epoch = 1, round = 16, global round = 16), Loss/Training: 1.6335580602288247
(epoch = 1, round = 16, global round = 16), Accuracy/Training: 40.6
reporting (epoch = 1, round = 16, global round = 16) for aggregation


Round:  80%|████████  | 16/20 [00:18<00:04,  1.19s/round]

(epoch = 1, round = 16, global round = 16), Loss/Aggregation: 1.5641359239816666
(epoch = 1, round = 16, global round = 16), Accuracy/Aggregation: 43.92
(epoch = 1, round = 16, global round = 16), Client to Server Bytes Sent/Training: 11032.0
(epoch = 1, round = 16, global round = 16), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 17
(epoch = 1, round = 17, global round = 17), Loss/Training: 1.6392898231744766
(epoch = 1, round = 17, global round = 17), Accuracy/Training: 40.04
reporting (epoch = 1, round = 17, global round = 17) for aggregation


Round:  85%|████████▌ | 17/20 [00:19<00:03,  1.17s/round]

(epoch = 1, round = 17, global round = 17), Loss/Aggregation: 1.5675200298428535
(epoch = 1, round = 17, global round = 17), Accuracy/Aggregation: 43.76
(epoch = 1, round = 17, global round = 17), Client to Server Bytes Sent/Training: 11032.0
(epoch = 1, round = 17, global round = 17), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 18
(epoch = 1, round = 18, global round = 18), Loss/Training: 1.600716432929039
(epoch = 1, round = 18, global round = 18), Accuracy/Training: 43.0
reporting (epoch = 1, round = 18, global round = 18) for aggregation


Round:  90%|█████████ | 18/20 [00:20<00:02,  1.16s/round]

(epoch = 1, round = 18, global round = 18), Loss/Aggregation: 1.5557768046855927
(epoch = 1, round = 18, global round = 18), Accuracy/Aggregation: 44.44
(epoch = 1, round = 18, global round = 18), Client to Server Bytes Sent/Training: 11032.0
(epoch = 1, round = 18, global round = 18), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 19
(epoch = 1, round = 19, global round = 19), Loss/Training: 1.5778911039233208
(epoch = 1, round = 19, global round = 19), Accuracy/Training: 42.56
reporting (epoch = 1, round = 19, global round = 19) for aggregation


Round:  95%|█████████▌| 19/20 [00:21<00:01,  1.14s/round]

(epoch = 1, round = 19, global round = 19), Loss/Aggregation: 1.522291925549507
(epoch = 1, round = 19, global round = 19), Accuracy/Aggregation: 44.84
(epoch = 1, round = 19, global round = 19), Client to Server Bytes Sent/Training: 11032.0
(epoch = 1, round = 19, global round = 19), Server to Client Bytes Sent/Training: 126376.0
Train finished Global Round: 20
(epoch = 1, round = 20, global round = 20), Loss/Training: 1.5901262491941452
(epoch = 1, round = 20, global round = 20), Accuracy/Training: 42.08
reporting (epoch = 1, round = 20, global round = 20) for aggregation
(epoch = 1, round = 20, global round = 20), Loss/Aggregation: 1.5773373901844026
(epoch = 1, round = 20, global round = 20), Accuracy/Aggregation: 43.36
(epoch = 1, round = 20, global round = 20), Client to Server Bytes Sent/Training: 11032.0
(epoch = 1, round = 20, global round = 20), Server to Client Bytes Sent/Training: 126376.0
Running (epoch = 1, round = 20, global round = 20) for Eval
(epoch = 1, round = 20, g

Epoch:   0%|          | 0/1 [00:23<?, ?epoch/s]


This tutorial has provided a concise overview of how to use the existing channel features in FLSim for communication efficiency. This should be useful for any development work on experimenting with different compression approaches to enable scalable federated learning. Please refer to https://github.com/facebookresearch/FLSim/tree/main/channels for the code pointers to implementation of different channels as well as the communication metrics reporting.